src/main/java/io/cdap/plugin/couchbase/source/CouchbaseSource.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.plugin.couchbase.source;
+
+import com.couchbase.client.java.document.json.JsonObject;
+import com.couchbase.client.java.query.N1qlQueryRow;
+import com.google.common.base.Preconditions;
+import io.cdap.cdap.api.annotation.Description;
+import io.cdap.cdap.api.annotation.Name;
+import io.cdap.cdap.api.annotation.Plugin;
+import io.cdap.cdap.api.data.batch.Input;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.dataset.lib.KeyValue;
+import io.cdap.cdap.etl.api.Emitter;
+import io.cdap.cdap.etl.api.FailureCollector;
+import io.cdap.cdap.etl.api.PipelineConfigurer;
+import io.cdap.cdap.etl.api.StageConfigurer;
+import io.cdap.cdap.etl.api.batch.BatchRuntimeContext;
+import io.cdap.cdap.etl.api.batch.BatchSource;
+import io.cdap.cdap.etl.api.batch.BatchSourceContext;
+import io.cdap.plugin.common.LineageRecorder;
+import io.cdap.plugin.couchbase.CouchbaseConfig;
+import io.cdap.plugin.couchbase.CouchbaseConstants;
+import org.apache.hadoop.io.NullWritable;
+
+import java.util.stream.Collectors;
+
+/**
+ * Plugin returns records from Couchbase Server.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name(CouchbaseConstants.PLUGIN_NAME)
+@Description("Read data from Couchbase Server.")
+public class CouchbaseSource extends BatchSource<NullWritable, N1qlQueryRow, StructuredRecord> {
+
+  private final CouchbaseConfig config;
+  private JsonObjectToRecordTransformer transformer;
+
+  public CouchbaseSource(CouchbaseConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    FailureCollector collector = stageConfigurer.getFailureCollector();
+    config.validate(collector);
```

Comment:
missed collector check?

Reply:
`collector.getOrThrowException();` was called by the config. Moved `getOrThrowException` invocation from the config to the source.